File: chrome-plugin/src/background/repair-errors.test.ts
import { describe, expect, it } from "vitest";
import { groupRepairErrors } from "./repair-errors";
import type { SentenceInput } from "../shared/protocol";
import { tokenize } from "../language/segmenter";
import type { ValidationError } from "../language/analysis-validator";

const sentence = (id: string): SentenceInput => ({
  sentenceId: id,
  text: "The service works well.",
  tokens: tokenize("The service works well."),
});

const error = (path: string, message: string): ValidationError => ({ path, message });

describe("groupRepairErrors", () => {
  it("多句 repair 的错误各自归属到 sentenceId,不再共享 sentences[0]", () => {
    const groups = groupRepairErrors(
      [
        {
          sentence: sentence("a"),
          errors: [error("sentences[0].components[1].translation", "m-a")],
        },
        {
          sentence: sentence("b"),
          errors: [error("sentences[0].components[3].translation", "m-b")],
        },
      ],
      { sentences: [{ sentenceId: "a" }, { sentenceId: "b" }] },
    );
    expect(groups).toEqual([
      {
        sentenceId: "a",
        rawOccurrence: 0,
        kind: "invalid",
        errors: [{ path: "components[1].translation", message: "m-a" }],
      },
      {
        sentenceId: "b",
        rawOccurrence: 0,
        kind: "invalid",
        errors: [{ path: "components[3].translation", message: "m-b" }],
      },
    ]);
  });

  it("缺失句与重复句各有明确 kind", () => {
    const missing = groupRepairErrors([{ sentence: sentence("a"), errors: [] }], { sentences: [] });
    expect(missing[0]).toMatchObject({ sentenceId: "a", kind: "missing" });

    const duplicate = groupRepairErrors(
      [
        {
          sentence: sentence("a"),
          errors: [error("sentences[1].sentenceId", "is duplicated")],
        },
      ],
      { sentences: [{ sentenceId: "a" }, { sentenceId: "a" }] },
    );
    expect(duplicate).toEqual([
      {
        sentenceId: "a",
        rawOccurrence: 1,
        kind: "duplicate",
        errors: [{ path: "", message: "remove this duplicate instance" }],
      },
    ]);
  });

  it("重复实例与首实例非法并存时,两边各出一组、互不吞并", () => {
    const groups = groupRepairErrors(
      [
        {
          sentence: sentence("a"),
          errors: [
            error("sentences[1].sentenceId", "is duplicated"),
            error("sentences[0].components[1].translation", "m-a"),
          ],
        },
      ],
      { sentences: [{ sentenceId: "a" }, { sentenceId: "a" }, { sentenceId: "a" }] },
    );
    expect(groups).toEqual([
      {
        sentenceId: "a",
        rawOccurrence: 0,
        kind: "invalid",
        errors: [{ path: "components[1].translation", message: "m-a" }],
      },
      {
        sentenceId: "a",
        rawOccurrence: 1,
        kind: "duplicate",
        errors: [{ path: "", message: "remove this duplicate instance" }],
      },
      {
        sentenceId: "a",
        rawOccurrence: 2,
        kind: "duplicate",
        errors: [{ path: "", message: "remove this duplicate instance" }],
      },
    ]);
  });

  it("句对象级错误去掉 sentences 前缀后保留句级根路径", () => {
    const groups = groupRepairErrors(
      [{ sentence: sentence("a"), errors: [error("sentences[0].sentenceId", "must be a safe string")] }],
      { sentences: [{ sentenceId: "a" }] },
    );
    expect(groups).toEqual([
      {
        sentenceId: "a",
        rawOccurrence: 0,
        kind: "invalid",
        errors: [{ path: "sentenceId", message: "must be a safe string" }],
      },
    ]);
  });
});

File: intellij-plugin/src/test/kotlin/dev/codetui/englishsyntax/analysis/RepairErrorsTest.kt
package dev.codetui.englishsyntax.analysis

import dev.codetui.englishsyntax.domain.SentenceInput
import dev.codetui.englishsyntax.domain.Token
import dev.codetui.englishsyntax.domain.ValidationError
import dev.codetui.englishsyntax.language.tokenize
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlin.test.Test
import kotlin.test.assertEquals

/**
 * 与 TS `repair-errors.test.ts` 同构:按句分组、四类路径、重复与非法并存。
 */
class RepairErrorsTest {
  private fun sentence(id: String): SentenceInput {
    val text = "The service works well."
    return SentenceInput(id, text, tokenize(text))
  }

  private fun raw(json: String): JsonElement = Json.parseToJsonElement(json)

  @Test
  fun `多句错误各自归属到 sentenceId 且路径去掉 sentences 前缀`() {
    val groups = groupRepairErrors(
      listOf(
        sentence("a") to listOf(ValidationError("sentences[0].components[1].translation", "m-a")),
        sentence("b") to listOf(ValidationError("sentences[0].components[3].translation", "m-b")),
      ),
      raw("""{"sentences":[{"sentenceId":"a"},{"sentenceId":"b"}]}"""),
    )
    assertEquals(
      listOf(
        RepairErrorGroup("a", 0, "invalid", listOf(ValidationError("components[1].translation", "m-a"))),
        RepairErrorGroup("b", 0, "invalid", listOf(ValidationError("components[3].translation", "m-b"))),
      ),
      groups,
    )
  }

  @Test
  fun `缺失句与重复句各有明确 kind`() {
    val missing = groupRepairErrors(
      listOf(sentence("a") to emptyList()),
      raw("""{"sentences":[]}"""),
    )
    assertEquals(RepairErrorGroup("a", null, "missing", listOf(ValidationError("", "no output for this sentenceId; emit it"))), missing.single())

    val duplicate = groupRepairErrors(
      listOf(sentence("a") to listOf(ValidationError("sentences[1].sentenceId", "is duplicated"))),
      raw("""{"sentences":[{"sentenceId":"a"},{"sentenceId":"a"}]}"""),
    )
    assertEquals(
      listOf(RepairErrorGroup("a", 1, "duplicate", listOf(ValidationError("", "remove this duplicate instance")))),
      duplicate,
    )
  }

  @Test
  fun `重复实例与首实例非法并存时互不吞并`() {
    val groups = groupRepairErrors(
      listOf(
        sentence("a") to listOf(
          ValidationError("sentences[1].sentenceId", "is duplicated"),
          ValidationError("sentences[0].components[1].translation", "m-a"),
        ),
      ),
      raw("""{"sentences":[{"sentenceId":"a"},{"sentenceId":"a"},{"sentenceId":"a"}]}"""),
    )
    assertEquals(
      listOf(
        RepairErrorGroup("a", 0, "invalid", listOf(ValidationError("components[1].translation", "m-a"))),
        RepairErrorGroup("a", 1, "duplicate", listOf(ValidationError("", "remove this duplicate instance"))),
        RepairErrorGroup("a", 2, "duplicate", listOf(ValidationError("", "remove this duplicate instance"))),
      ),
      groups,
    )
  }
}
